ath79: add support for MikroTik RouterBOARD 750r2 (hEX lite)

rogerpueyo opened the pull request:
This patch adds support for the MikroTik RouterBOARD 750r2, marketed as
hEX lite, a small indoor router with 5x 10/100 Mbps Ethernet ports, one
with PoE in. The device was already supported by the ar71xx target.

Specifications:
 - SoC: Qualcomm Atheros QCA9533
 - Flash: 16 MB SPI NOR
 - RAM: 64 MB
 - Ethernet: 4x 10/100 Mbps LAN, 1x 10/100 Mbps WAN (PoE in)
 - LEDs: 5x Ethernet port activity (green), 1x user (green)
 - Buttons: 1x reset

 See https://mikrotik.com/product/RB750r2 for more details.

Not working:
 - Serial port (already not working in ar71xx)

Flashing:
 TFTP boot initramfs image and then perform sysupgrade. Only the
 "Internet" port will ask for an initramfs image. Follow common
 MikroTik procedure as in https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Roger Pueyo Centelles <roger.pueyo@guifi.net>

rogerpueyo committed:
ath79: add support for MikroTik RouterBOARD 750 r2 (hEX lite)

This patch adds support for the MikroTik RouterBOARD 750 r2, marketed as
hEX lite, a small indoor router with 5x 10/100 Mbps Ethernet ports, one
with PoE in. The device was already supported by the ar71xx target.

Specifications:
 - SoC: Qualcomm Atheros QCA9533
 - Flash: 16 MB SPI NOR
 - RAM: 64 MB
 - Ethernet: 4x 10/100 Mbps LAN, 1x 10/100 Mbps WAN (PoE in)
 - LEDs: 5x Ethernet port activity (green), 1x user (green)
 - Buttons: 1x reset

 See https://mikrotik.com/product/RB750r2 for more details.

Not working:
 - Serial port (already not working in ar71xx)

Flashing:
 TFTP boot initramfs image and then perform sysupgrade. Only the
 "Internet" port will ask for an initramfs image. Follow common
 MikroTik procedure as in https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Roger Pueyo Centelles <roger.pueyo@guifi.net>

rogerpueyo commented:
BTW, the ath79/mikrotik subtarget is affected by https://bugs.openwrt.org/index.php?do=details&task_id=3492

rogerpueyo commented:
(rebased master)

adschm commented on target/linux/ath79/mikrotik/base-files/etc/board.d/02_network:
```diff
@@ -14,6 +14,11 @@ ath79_setup_interfaces()
 			"0@eth1" "1:lan:4" "2:lan:1" "3:lan:2" "4:lan:3"
 		;;
+	mikrotik,routerboard-750-r2)
+		ucidef_set_interfaces_lan_wan "eth0.1" "eth1"
```
ucidef_set_interface_wan "eth1"
should be enough, lan is set from the switch command.

rogerpueyo replied:
True, it's redundant. Thanks.

adschm commented on target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds:
```diff
@@ -6,6 +6,13 @@ board_config_update
+	ucidef_set_led_switch "led2" "2" "green:led2" "switch0" "0x10"
+	ucidef_set_led_switch "led3" "3" "green:led3" "switch0" "0x08"
+	ucidef_set_led_switch "led4" "4" "green:led4" "switch0" "0x04"
+	ucidef_set_led_switch "led5" "5" "green:led5" "switch0" "0x02"
```
1. Is this vendor behavior?
2. Does this really match the port assignment in 02_network?

adschm commented on target/linux/ath79/dts/qca9533_mikrotik_routerboard-750-r2.dts:
```diff
@@ -0,0 +1,83 @@
+};
+
+&spi {
+
```
drop empty line.

rogerpueyo replied:
Removed. Thanks.

adschm commented on target/linux/ath79/dts/qca9533_mikrotik_routerboard-750-r2.dts:
```diff
@@ -0,0 +1,83 @@
+
+&eth1 {
+	status = "okay";
+};
```
status is already set to okay in qca953x.dtsi for eth1. This block can be dropped.

rogerpueyo replied:
True. Thanks.

adschm commented on target/linux/ath79/mikrotik/base-files/etc/board.d/02_network:
```diff
@@ -14,6 +14,11 @@ ath79_setup_interfaces()
+		ucidef_set_interfaces_lan_wan "eth0.1" "eth1"
+		ucidef_add_switch "switch0" \
+			"0@eth0" "1:lan:4" "2:lan:1" "3:lan:3" "4:lan:2"
+		;;
```
MAC assignment missing at the bottom of this file?

adschm replied:
I forgot that mikrotik subtarget is special and has a default case here. Never mind ...

rogerpueyo commented:
Force-pushed with changes requested by @adschm and rebase from master.

rogerpueyo commented:
Changes requested at review time were made. Force-pushed with rebase from master.

goosnarrggh commented:
This patch is targeted at master, which means that if successful it would eventually end up in a future 22.x (or higher) release. Would it also be possible for an individual to apply it to the 21.02.x branch? (I've noticed that the base files which this patch is applied against also seem to be present in 21.02.x.)

Also, is there anything which a regular user, comfortable with building from source code but without any specific experience contributing to OpenWRT, might be able to do to help move along the review process?

rogerpueyo commented:
> This patch is targeted at master, which means that if successful it would eventually end up in a future 22.x (or higher) release. Would it also be possible for an individual to apply it to the 21.02.x branch? (I've noticed that the base files which this patch is applied against also seem to be present in 21.02.x.)

Yes, this is the usual procedure. Unfortunately, it doesn't seem it'll go into 22.x neither, since it's been already branched.
 
> Also, is there anything which a regular user, comfortable with building from source code but without any specific experience contributing to OpenWRT, might be able to do to help move along the review process?

You could clone the repository, build the image, and report if everything works fine (network, settings, LEDs, etc.).

I rebased current master.

hauke commented on target/linux/ath79/dts/qca9533_mikrotik_routerboard-wapr-2nd.dts:
```diff
@@ -50,6 +50,12 @@
+&wmac {
+	status = "okay";
+
+	qca,no-eeprom;
```
This is now included in more files. I think you have to activate this in more files. 
````
target/linux/ath79/dts/qca9533_mikrotik_routerboard-lhg-hb.dtsi:#include "qca9533_mikrotik_routerboard-16m.dtsi"
target/linux/ath79/dts/qca9533_mikrotik_routerboard-mapl-2nd.dts:#include "qca9533_mikrotik_routerboard-16m.dtsi"
target/linux/ath79/dts/qca9533_mikrotik_routerboard-wap-2nd.dts:#include "qca9533_mikrotik_routerboard-16m.dtsi"
target/linux/ath79/dts/qca9533_mikrotik_routerboard-wapr-2nd.dts:#include "qca9533_mikrotik_routerboard-16m.dtsi"
```

rogerpueyo replied:
Damn, I didn't think of new devices being added. Thanks for spotting this! I'm fixing it and rebasing.

rogerpueyo replied:
Thanks for your review, @hauke.

I added the removed `&wmac` element to the wireless-capable new devices supported after the rebase.

WolleTD commented:
So, this apparently doesn't make it into 23.x either?

For the record: I have successfully built, netbooted, flashed and flash-booted this configuration, cherry-picked onto 23.x branch.

hauke commented:
There are now even more boards affected, see:
````
target/linux/ath79/dts/qca9533_mikrotik_routerboard-95x.dtsi:#include "qca9533_mikrotik_routerboard-16m.dtsi"
target/linux/ath79/dts/qca9533_mikrotik_routerboard-lhg-hb.dtsi:#include "qca9533_mikrotik_routerboard-16m.dtsi"
target/linux/ath79/dts/qca9533_mikrotik_routerboard-map-2nd.dts:#include "qca9533_mikrotik_routerboard-16m.dtsi"
target/linux/ath79/dts/qca9533_mikrotik_routerboard-mapl-2nd.dts:#include "qca9533_mikrotik_routerboard-16m.dtsi"
target/linux/ath79/dts/qca9533_mikrotik_routerboard-wap-2nd.dts:#include "qca9533_mikrotik_routerboard-16m.dtsi"
target/linux/ath79/dts/qca9533_mikrotik_routerboard-wapr-2nd.dts:#include "qca9533_mikrotik_routerboard-16m.dtsi"
````

Please also rebase on top of current master.

WolleTD commented:
Not the original author, but the rebased and fixed commit could be found here: https://github.com/WolleTD/openwrt/commit/9e0921382d3c08c663f3b044a259adc9f6b537dc

rogerpueyo commented:
Hi,

I just rebased current master. Sorry, I couldn't do that before. Should be the same as WolleTD@9e09213 except for the LED names, which I changed from `ledX` to `portX` so that they match with other devices already supported.

I'll be testing it during the afternoon.

WolleTD commented:
> ~I'll be testing it during the afternoon.~ Edit: tested on the device, working.

I also successfully built and tested the current PR commit:
> Firmware Version | OpenWrt SNAPSHOT r0+23916-09c2f8228d

hauke commented:
Sorry for the long delay to get this merged.

WolleTD commented:
Cool, thanks! As I've seen other device support commits in the 23.05 branch, is it possible this will get official support with 23.05 as well?

rogerpueyo commented:
> Cool, thanks! As I've seen other device support commits in the 23.05 branch, is it possible this will get official support with 23.05 as well?

Since it's a fairly trivial device support, we might be lucky to see it in 23.05: https://github.com/openwrt/openwrt/pull/13477
